File: CS/E1271/Record.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Text;

namespace E1271 {
    public class Record : INotifyPropertyChanged {
        public Record() {
            Details = new BindingList<Detail>();
        }
        int id;
        public int ID {
            get { return id; }
            set {
                if(id != value) {
                    id = value;
                    OnPropertyChanged("ID");
                }
            }
        }
        int parentID;
        public int ParentID {
            get { return parentID; }
            set {
                if(parentID != value) {
                    parentID = value;
                    OnPropertyChanged("ParentID");
                }
            }
        }
        string text;
        public string Text {
            get { return text; }
            set {
                if(text != value) {
                    text = value;
                    OnPropertyChanged("Text");
                }
            }
        }
        string info;
        public string Info {
            get { return info; }
            set {
                if(info != value) {
                    info = value;
                    OnPropertyChanged("Info");
                }
            }
        }
        decimal? val;
        public decimal? Value {
            get { return val; }
            set {
                if(val != value) {
                    val = value;
                    OnPropertyChanged("Value");
                }
            }
        }
        DateTime dt;
        public DateTime Dt {
            get { return dt; }
            set {
                if(dt != value) {
                    dt = value;
                    OnPropertyChanged("Dt");
                }
            }
        }
        bool state;
        public bool State {
            get { return state; }
            set {
                if(state != value) {
                    state = value;
                    OnPropertyChanged("State");
                }
            }
        }
        Image image;
        public Image Image {
            get { return image; }
            set {
                if(image != value) {
                    image = value;
                    OnPropertyChanged("Image");
                }
            }
        }
        BindingList<Detail> details;
        public BindingList<Detail> Details {
            get { return details; }
            set {
                if(details != value) {
                    details = value;
                    OnPropertyChanged("Details");
                }
            }
        }
        public override string ToString() {
            return string.Format("ID = {0}", ID, Text);
        }
        Detail detail;
        public Detail Detail {
            get { return detail; }
            set {
                if(detail != value) {
                    detail = value;
                    OnPropertyChanged("Detail");
                }
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;
        protected void OnPropertyChanged(string propertyName) {
            if(PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
    public class Detail : INotifyPropertyChanged {
        public Detail() {
        }
        int id;
        public int ID {
            get { return id; }
            set {
                if(id != value) {
                    id = value;
                    OnPropertyChanged("ID");
                }
            }
        }
        string name;
        public string Name {
            get { return name; }
            set {
                if(name != value) {
                    name = value;
                    OnPropertyChanged("Name");
                }
            }
        }
        string info;
        public string Info {
            get { return info; }
            set {
                if(info != value) {
                    info = value;
                    OnPropertyChanged("Info");
                }
            }
        }
        bool state;
        public bool State {
            get { return state; }
            set {
                if(state != value) {
                    state = value;
                    OnPropertyChanged("State");
                }
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;
        protected void OnPropertyChanged(string propertyName) {
            if(PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
    public class DataHelper {
        public static BindingList<Record> GetData(int count) {
            BindingList<Record> records = new BindingList<Record>();
            for(int i = 0; i < count; i++) {
                records.Add(new Record() {
                    ParentID = i < 5 ? 1 : 2,
                    ID = i,
                    Text = string.Format("Text {0}", i),
                    Info = string.Format("Info {0} Line 1\r\n Line 2\r\n Line 3\r\n Line 4\r\n Line 5\r\n", i),
                    Dt = i % 2 == 0 ? DateTime.Now.AddDays(i) : DateTime.MinValue,
                    State = i % 2 == 0,
                    Image = SystemIcons.Information.ToBitmap(),
                    Value = (i * 0.00001M),
                    Detail = new Detail() { ID = i, Name = string.Format("Text {0}", i), State = i % 2 == 0 }
                });
            }
            return records;
        }
        public static BindingList<Record> GetData(int count, int detailCount) {
            BindingList<Record> records = new BindingList<Record>();
            for(int i = 0; i < count; i++) {
                records.Add(new Record() {
                    ParentID = i < 5 ? 1 : 2,
                    ID = i,
                    Text = string.Format("Text {0}", i),
                    Info = string.Format("Info {0} Line 1\r\n Line 2\r\n Line 3\r\n Line 4\r\n Line 5\r\n", i),
                    Dt = i % 2 == 0 ? DateTime.Now.AddDays(i) : DateTime.MinValue,
                    State = i % 2 == 0,
                    Image = SystemIcons.Information.ToBitmap(),
                    Value = (i * 0.00001M),
                    Detail = new Detail() { ID = i, Name = string.Format("Text {0}", i), State = i % 2 == 0 }
                });
                BindingList<Detail> details = new BindingList<Detail>();
                for(int j = 0; j < detailCount; j++) {
                    details.Add(new Detail() {
                        ID = j,
                        Name = string.Format("Text {0}", j),
                        Info = string.Format("Info {0}", j)
                    });
                }
                records[i].Details = details;
            }

            return records;
        }
    }
}

File: VB/E1271/Record.vb
Imports System.ComponentModel
Imports System.Drawing

Namespace E1271

    Public Class Record
        Implements INotifyPropertyChanged

        Public Sub New()
            Details = New BindingList(Of Detail)()
        End Sub

        Private idField As Integer

        Public Property ID As Integer
            Get
                Return idField
            End Get

            Set(ByVal value As Integer)
                If idField <> value Then
                    idField = value
                    OnPropertyChanged("ID")
                End If
            End Set
        End Property

        Private parentIDField As Integer

        Public Property ParentID As Integer
            Get
                Return parentIDField
            End Get

            Set(ByVal value As Integer)
                If parentIDField <> value Then
                    parentIDField = value
                    OnPropertyChanged("ParentID")
                End If
            End Set
        End Property

        Private textField As String

        Public Property Text As String
            Get
                Return textField
            End Get

            Set(ByVal value As String)
                If Not Equals(textField, value) Then
                    textField = value
                    OnPropertyChanged("Text")
                End If
            End Set
        End Property

        Private infoField As String

        Public Property Info As String
            Get
                Return infoField
            End Get

            Set(ByVal value As String)
                If Not Equals(infoField, value) Then
                    infoField = value
                    OnPropertyChanged("Info")
                End If
            End Set
        End Property

        Private val As Decimal?

        Public Property Value As Decimal?
            Get
                Return val
            End Get

            Set(ByVal value As Decimal?)
                If val <> value Then
                    val = value
                    OnPropertyChanged("Value")
                End If
            End Set
        End Property

        Private dtField As Date

        Public Property Dt As Date
            Get
                Return dtField
            End Get

            Set(ByVal value As Date)
                If dtField <> value Then
                    dtField = value
                    OnPropertyChanged("Dt")
                End If
            End Set
        End Property

        Private stateField As Boolean

        Public Property State As Boolean
            Get
                Return stateField
            End Get

            Set(ByVal value As Boolean)
                If stateField <> value Then
                    stateField = value
                    OnPropertyChanged("State")
                End If
            End Set
        End Property

        Private imageField As Image

        Public Property Image As Image
            Get
                Return imageField
            End Get

            Set(ByVal value As Image)
                If imageField IsNot value Then
                    imageField = value
                    OnPropertyChanged("Image")
                End If
            End Set
        End Property

        Private detailsField As BindingList(Of Detail)

        Public Property Details As BindingList(Of Detail)
            Get
                Return detailsField
            End Get

            Set(ByVal value As BindingList(Of Detail))
                If detailsField IsNot value Then
                    detailsField = value
                    OnPropertyChanged("Details")
                End If
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return String.Format("ID = {0}", ID, Text)
        End Function

        Private detailField As Detail

        Public Property Detail As Detail
            Get
                Return detailField
            End Get

            Set(ByVal value As Detail)
                If detailField IsNot value Then
                    detailField = value
                    OnPropertyChanged("Detail")
                End If
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Sub OnPropertyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class

    Public Class Detail
        Implements INotifyPropertyChanged

        Public Sub New()
        End Sub

        Private idField As Integer

        Public Property ID As Integer
            Get
                Return idField
            End Get

            Set(ByVal value As Integer)
                If idField <> value Then
                    idField = value
                    OnPropertyChanged("ID")
                End If
            End Set
        End Property

        Private nameField As String

        Public Property Name As String
            Get
                Return nameField
            End Get

            Set(ByVal value As String)
                If Not Equals(nameField, value) Then
                    nameField = value
                    OnPropertyChanged("Name")
                End If
            End Set
        End Property

        Private infoField As String

        Public Property Info As String
            Get
                Return infoField
            End Get

            Set(ByVal value As String)
                If Not Equals(infoField, value) Then
                    infoField = value
                    OnPropertyChanged("Info")
                End If
            End Set
        End Property

        Private stateField As Boolean

        Public Property State As Boolean
            Get
                Return stateField
            End Get

            Set(ByVal value As Boolean)
                If stateField <> value Then
                    stateField = value
                    OnPropertyChanged("State")
                End If
            End Set
        End Property

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Sub OnPropertyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class

    Public Class DataHelper

        Public Shared Function GetData(ByVal count As Integer) As BindingList(Of Record)
            Dim records As BindingList(Of Record) = New BindingList(Of Record)()
            For i As Integer = 0 To count - 1
                records.Add(New Record() With {.ParentID = If(i < 5, 1, 2), .ID = i, .Text = String.Format("Text {0}", i), .Info = String.Format("Info {0} Line 1" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 2" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 3" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 4" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 5" & Microsoft.VisualBasic.Constants.vbCrLf, i), .Dt = If(i Mod 2 = 0, Date.Now.AddDays(i), Date.MinValue), .State = i Mod 2 = 0, .Image = SystemIcons.Information.ToBitmap(), .Value = i * 0.00001D, .Detail = New Detail() With {.ID = i, .Name = String.Format("Text {0}", i), .State = i Mod 2 = 0}})
            Next

            Return records
        End Function

        Public Shared Function GetData(ByVal count As Integer, ByVal detailCount As Integer) As BindingList(Of Record)
            Dim records As BindingList(Of Record) = New BindingList(Of Record)()
            For i As Integer = 0 To count - 1
                records.Add(New Record() With {.ParentID = If(i < 5, 1, 2), .ID = i, .Text = String.Format("Text {0}", i), .Info = String.Format("Info {0} Line 1" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 2" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 3" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 4" & Microsoft.VisualBasic.Constants.vbCrLf & " Line 5" & Microsoft.VisualBasic.Constants.vbCrLf, i), .Dt = If(i Mod 2 = 0, Date.Now.AddDays(i), Date.MinValue), .State = i Mod 2 = 0, .Image = SystemIcons.Information.ToBitmap(), .Value = i * 0.00001D, .Detail = New Detail() With {.ID = i, .Name = String.Format("Text {0}", i), .State = i Mod 2 = 0}})
                Dim details As BindingList(Of Detail) = New BindingList(Of Detail)()
                For j As Integer = 0 To detailCount - 1
                    details.Add(New Detail() With {.ID = j, .Name = String.Format("Text {0}", j), .Info = String.Format("Info {0}", j)})
                Next

                records(i).Details = details
            Next

            Return records
        End Function
    End Class
End Namespace
